TombEngine/Objects/TR3/Entity/FishSwarm.cpp

```diff
@@ -0,0 +1,410 @@
+	constexpr auto FISH_BASE_SEPARATION_DISTANCE = 210.0f;
+
+	std::vector<FishData> FishSwarm = {};
+
+	void InitializeFishSwarm(short itemNumber)
+	{
+		constexpr auto DEFAULT_FISH_COUNT = 24;
+
+		auto& item = g_Level.Items[itemNumber];
+
+		item.StartPose.Position = item.Pose.Position;
+		item.Animation.Velocity.z = Random::GenerateFloat(32.0f, 160.0f);
+		item.HitPoints = DEFAULT_FISH_COUNT;
+		item.ItemFlags[0] = item.Index;
+		item.ItemFlags[1] = item.Index;
+		item.ItemFlags[5] = 0;
+
+		if (item.AIBits)
+			item.ItemFlags[6] = true;
+	}
+
+	static void SpawnFishSwarm(ItemInfo& item)
+	{
+		constexpr auto VEL_MAX				   = 48.0f;
+		constexpr auto VEL_MIN				   = 16.0f;
+		constexpr auto START_ORIENT_CONSTRAINT = std::pair<EulerAngles, EulerAngles>(
+			EulerAngles(ANGLE(-3.0f), ANGLE(-6.0f), 0),
+			EulerAngles(ANGLE(3.0f), ANGLE(6.0f), 0));
+
+		// Create new fish.
+		auto& fish = GetNewEffect(FishSwarm, FISH_COUNT_MAX);
+
+		fish.MeshIndex = abs(item.TriggerFlags);
+		fish.IsLethal = (item.TriggerFlags < 0) ? true : false;
+		fish.IsPatrolling = item.ItemFlags[6];
+
+		fish.Position = item.Pose.Position.ToVector3();
+		fish.RoomNumber = item.RoomNumber;
+		fish.Orientation.x = Random::GenerateAngle(START_ORIENT_CONSTRAINT.first.x, START_ORIENT_CONSTRAINT.second.x);
+		fish.Orientation.y = (item.Pose.Orientation.y + ANGLE(180.0f)) + Random::GenerateAngle(START_ORIENT_CONSTRAINT.first.y, START_ORIENT_CONSTRAINT.second.y);
+		fish.Velocity = Random::GenerateFloat(VEL_MIN, VEL_MAX);
+
+		fish.Life = 1.0f;
+		fish.Undulation = Random::GenerateFloat(0.0f, PI_MUL_2);
+
+		fish.LeaderItemPtr = &g_Level.Items[item.ItemFlags[0]];
+	}
+
+	void ControlFishSwarm(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+		auto& creature = *GetCreatureInfo(&item);
+
+		if (!CreatureActive(itemNumber))
+			return;
+
+		AI_INFO ai;
+		CreatureAIInfo(&item, &ai);
+
+		if (item.HitPoints != NOT_TARGETABLE)
+		{
+			int fishCount = item.HitPoints - item.ItemFlags[5];
+
+			if (fishCount < 0)
+			{
+				int fishToTurnOff = -fishCount;
+				for (auto& fish : FishSwarm)
+				{
+					if (fish.LeaderItemPtr == &item && fish.Life > 0.0f)
+					{
+						fish.Life = 0.0f;
+						fishToTurnOff--;
+						if (fishToTurnOff == 0)
+							break;
+					}
+				}
+			}
+			else if (fishCount > 0)
+			{
+				for (int i = 0; i < fishCount; i++)
+					SpawnFishSwarm(item);
+			}
+
+			item.ItemFlags[5] = item.HitPoints;
+			item.HitPoints = NOT_TARGETABLE;
+		}
+
+		int dx = creature.Target.x - item.Pose.Position.x;
+		int dz = creature.Target.z - item.Pose.Position.z;
+		ai.distance = SQUARE(dx) + SQUARE(dz);
+
+		item.Animation.Velocity.z = FISH_VELOCITY_MAX;
+
+		auto& playerRoom = g_Level.Rooms[LaraItem->RoomNumber];
+
+		auto corpsePos = std::optional<Vector3>();
+
+		// Check if corpse is near.
+		if (!corpsePos.has_value())
+		{
+			float closestDist = INFINITY;
+			for (auto& targetItem : g_Level.Items)
```

Comment:
Cycling through all items in a level every frame (and for every fish swarm) may become extremely slow. Can we consider narrowing search, for example, by searching only in neighbor rooms of a fish swarm? Additionally, this search may be done not every single frame, but e.g. every second or fourth frame (this can be achieved by using bitwise operation against `GlobalCounter` variable.